src/core/replicated_archetypes.rs

```diff
@@ -0,0 +1,103 @@
+use bevy::{
+    ecs::{
+        archetype::ArchetypeId,
+        component::{ComponentId, StorageType},
+    },
+    prelude::*,
+};
+
+use crate::core::{replication_fns::SerdeFnsId, Replication};
+
+/// Stores cached information about all replicated archetypes.
+///
+/// By default it's updated with [component-based](../../index.html#component-replication) replication rules.
+///
+/// But it's also possible to implement custom rules:
+/// - Register 'serde' and 'remove' functions inside [`ReplicationFns`](crate::core::replication_fns::ReplicationFns).
+/// - Update this struct for all newly added archetypes in
+/// [`ServerSet::UpdateArchetypes`](crate::server::ServerSet::UpdateArchetypes) using the registered function IDs.
+/// - Update [`RemovalBuffer`](crate::server::world_buffers::RemovalBuffer) in
+/// [`ServerSet::BufferRemovals`](crate::server::ServerSet::BufferRemovals) when the rule components should be removed.
```

Comment:
Right, we should mention it!

> note this implies a fully-custom implementation will have 50% empty archetypes in ReplicatedArchetypes

Hm... This could be a problem. Custom implementation could disable component-based rules, but it will mean that combining them will be inefficient.

What is your opinion about it?

Reply:
I like it!
Let me think how can I rework to make it work as you suggested...

Would you prefer that I open a separate PR to this branch or commit directly here?

Reply:
> since you may not know all relevant archetypes at startup

True, but I thought that you suggested to provide an API that doesn't operate on archetypes. And replicon will apply it to archetypes internally :sweat_smile: Anyway, I will explore this path. Not quite satisfied with the current solution.